pkg/providers/vault/provider.go

```diff
@@ -0,0 +1,370 @@
+	// Set vault token.
+	req.Header.Set("X-Vault-Token", token)
+	if err != nil {
+		return "", errors.Wrapf(err, "couldn't generate request")
+	}
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return "", errors.Wrapf(err, "couldn't get secret")
+	}
+
+	defer resp.Body.Close()
+
+	if resp.StatusCode != 200 {
+		var b bytes.Buffer
+		io.Copy(&b, resp.Body)
+		return "", fmt.Errorf("failed to get successful response: %#v, %s",
+			resp, b.String())
+	}
+
+	var d struct {
+		Data struct {
+			Data map[string]string `json:"data"`
+		} `json:"data"`
+	}
+
+	if err := json.NewDecoder(resp.Body).Decode(&d); err != nil {
+		return "", errors.Wrapf(err, "failed to read body")
+	}
+
+	return d.Data.Data[secretName], nil
+}
+
+func (p *Provider) getRootCAsPools() (*x509.CertPool, error) {
+	switch {
+	case p.VaultCAPem != "":
+		certPool := x509.NewCertPool()
+		if err := loadCert(certPool, []byte(p.VaultCAPem)); err != nil {
+			return nil, err
+		}
+		return certPool, nil
+	case p.VaultCAPath != "":
+		certPool := x509.NewCertPool()
+		if err := loadCertFolder(certPool, p.VaultCAPath); err != nil {
+			return nil, err
+		}
+		return certPool, nil
+	case p.VaultCACert != "":
+		certPool := x509.NewCertPool()
+		if err := loadCertFile(certPool, p.VaultCACert); err != nil {
+			return nil, err
+		}
+		return certPool, nil
+	default:
+		certPool, err := x509.SystemCertPool()
+		if err != nil {
+			return nil, errors.Wrapf(err, "couldn't load system certs")
+		}
+		return certPool, err
+	}
+}
+
+// loadCert loads a single pem-encoded certificate into the given pool.
+func loadCert(pool *x509.CertPool, pem []byte) error {
+	if ok := pool.AppendCertsFromPEM(pem); !ok {
+		return fmt.Errorf("failed to parse PEM")
+	}
+	return nil
+}
+
+// loadCertFile loads the certificate at the given path into the given pool.
+func loadCertFile(pool *x509.CertPool, p string) error {
+	pem, err := ioutil.ReadFile(p)
+	if err != nil {
+		return errors.Wrapf(err, "couldn't read CA file from disk")
+	}
+
+	if err := loadCert(pool, pem); err != nil {
+		return errors.Wrapf(err, "couldn't load CA at %s", p)
+	}
+
+	return nil
+}
+
+// loadCertFolder iterates exactly one level below the given directory path and
+// loads all certificates in that path. It does not recurse
+func loadCertFolder(pool *x509.CertPool, p string) error {
+	if err := filepath.Walk(p, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+
+		if info.IsDir() {
+			return nil
+		}
+
+		return loadCertFile(pool, path)
+	}); err != nil {
+		return errors.Wrapf(err, "failed to load CAs at %s", p)
+	}
+	return nil
+}
+
+// MountSecretsStoreObjectContent mounts content of the vault object to target path
+func (p *Provider) MountSecretsStoreObjectContent(ctx context.Context, attrib map[string]string, secrets map[string]string, targetPath string, permission os.FileMode) (err error) {
+	roleName := attrib["roleName"]
+	if roleName != "" {
+		p.VaultRole = roleName
+	}
```

Comment:
Do we need to validate `roleName` parameter has some value?

Reply:
yep, we should ensure it is set.